File: ui/src/core/utils/paths.ts
/**
 * Standardized Filesystem Paths
 * Centralized path constants for consistent filesystem structure across frontend
 *
 * This mirrors the kernel and backend path structure to ensure consistency.
 * Any changes should be synchronized with:
 * - kernel/src/vfs/paths.rs
 * - backend/internal/shared/paths/paths.go
 */

// Mount points
export const MOUNTS = {
  STORAGE: '/storage',
  TMP: '/tmp',
  CACHE: '/cache',
} as const;

// Storage subdirectories
export const STORAGE = {
  /** Prebuilt applications bundled with the OS */
  NATIVE_APPS: '/storage/native-apps',

  /** User-generated or AI-generated applications */
  APPS: '/storage/apps',

  /** User files and data */
  USER: '/storage/user',

  /** System configuration and data */
  SYSTEM: '/storage/system',

  /** Shared libraries and resources */
  LIB: '/storage/lib',
} as const;

// User subdirectories
export const USER = {
  DOCUMENTS: '/storage/user/documents',
  DOWNLOADS: '/storage/user/downloads',
  PROJECTS: '/storage/user/projects',
} as const;

/**
 * Get application-specific paths
 */
export class AppPaths {
  constructor(private readonly appId: string) {
    if (!appId) {
      throw new Error('App ID cannot be empty');
    }
  }

  /** Get app's data directory */
  dataDir(): string {
    return `${STORAGE.APPS}/${this.appId}/data`;
  }

  /** Get app's config directory */
  configDir(): string {
    return `${STORAGE.APPS}/${this.appId}/config`;
  }

  /** Get app's cache directory */
  cacheDir(): string {
    return `${MOUNTS.CACHE}/${this.appId}`;
  }

  /** Get app's temp directory */
  tempDir(): string {
    return `${MOUNTS.TMP}/${this.appId}`;
  }

  /** Resolve a relative path within the app */
  resolve(relativePath: string): string {
    return `${STORAGE.APPS}/${this.appId}/${relativePath}`;
  }
}

/**
 * Create app-specific path helper
 */
export function appPath(appId: string): AppPaths {
  return new AppPaths(appId);
}

/**
 * Check if path is within allowed userspace
 */
export function isUserspacePath(path: string): boolean {
  return (
    path.startsWith(STORAGE.APPS) ||
    path.startsWith(STORAGE.USER) ||
    path.startsWith(MOUNTS.TMP) ||
    path.startsWith(MOUNTS.CACHE)
  );
}

/**
 * Check if path is a native app
 */
export function isNativeAppPath(path: string): boolean {
  return path.startsWith(STORAGE.NATIVE_APPS);
}

/**
 * Check if path is system-protected
 */
export function isSystemPath(path: string): boolean {
  return path.startsWith(STORAGE.SYSTEM);
}

/**
 * Normalize path (remove trailing slashes, clean up)
 */
export function normalizePath(path: string): string {
  // Remove trailing slash except for root
  if (path.length > 1 && path.endsWith('/')) {
    path = path.slice(0, -1);
  }

  // Remove duplicate slashes
  path = path.replace(/\/+/g, '/');

  return path;
}

/**
 * Join path segments
 */
export function joinPath(...segments: string[]): string {
  return normalizePath(segments.join('/'));
}

/**
 * Get parent directory of path
 */
export function dirname(path: string): string {
  const normalized = normalizePath(path);
  const lastSlash = normalized.lastIndexOf('/');

  if (lastSlash === 0) {
    return '/';
  }

  if (lastSlash === -1) {
    return '.';
  }

  return normalized.slice(0, lastSlash);
}

/**
 * Get filename from path
 */
export function basename(path: string): string {
  const normalized = normalizePath(path);
  const lastSlash = normalized.lastIndexOf('/');

  if (lastSlash === -1) {
    return normalized;
  }

  return normalized.slice(lastSlash + 1);
}

/**
 * All standard directories
 */
export const STANDARD_DIRECTORIES = [
  STORAGE.NATIVE_APPS,
  STORAGE.APPS,
  STORAGE.USER,
  STORAGE.SYSTEM,
  STORAGE.LIB,
  USER.DOCUMENTS,
  USER.DOWNLOADS,
  USER.PROJECTS,
] as const;

/**
 * Path validation
 */
export function validatePath(path: string): { valid: boolean; error?: string } {
  if (!path) {
    return { valid: false, error: 'Path cannot be empty' };
  }

  if (path.includes('..')) {
    return { valid: false, error: 'Path cannot contain ..' };
  }

  if (path.includes('//')) {
    return { valid: false, error: 'Path cannot contain //' };
  }

  return { valid: true };
}

File: kernel/src/vfs/paths.rs
/*!
 * Standard Filesystem Paths
 * Centralized path constants for consistent filesystem structure
 */

use std::path::{Path, PathBuf};

/// Standard mount points
pub mod mounts {
    pub const STORAGE: &str = "/storage";
    pub const TMP: &str = "/tmp";
    pub const CACHE: &str = "/cache";
}

/// Standard directories under /storage
pub mod storage {
    /// Native applications (prebuilt, bundled with OS)
    pub const NATIVE_APPS: &str = "/storage/native-apps";

    /// User-generated/AI-generated applications
    pub const APPS: &str = "/storage/apps";

    /// User files and data
    pub const USER: &str = "/storage/user";

    /// System configuration and data
    pub const SYSTEM: &str = "/storage/system";

    /// Shared libraries and resources
    pub const LIB: &str = "/storage/lib";
}

/// Application-specific paths
pub mod app {
    use std::path::PathBuf;
    use super::storage;

    /// Get app data directory
    pub fn data_dir(app_id: &str) -> PathBuf {
        PathBuf::from(storage::APPS).join(app_id).join("data")
    }

    /// Get app config directory
    pub fn config_dir(app_id: &str) -> PathBuf {
        PathBuf::from(storage::APPS).join(app_id).join("config")
    }

    /// Get app cache directory
    pub fn cache_dir(app_id: &str) -> PathBuf {
        PathBuf::from("/cache").join(app_id)
    }

    /// Get app temp directory
    pub fn temp_dir(app_id: &str) -> PathBuf {
        PathBuf::from("/tmp").join(app_id)
    }
}

/// User-specific paths
pub mod user {
    pub const DOCUMENTS: &str = "/storage/user/documents";
    pub const DOWNLOADS: &str = "/storage/user/downloads";
    pub const PROJECTS: &str = "/storage/user/projects";
}

/// All standard directories that should be created at init
pub fn standard_directories() -> Vec<&'static str> {
    vec![
        // Storage directories
        storage::NATIVE_APPS,
        storage::APPS,
        storage::USER,
        storage::SYSTEM,
        storage::LIB,

        // User directories
        user::DOCUMENTS,
        user::DOWNLOADS,
        user::PROJECTS,
    ]
}

/// Check if path is within allowed userspace
pub fn is_userspace_path(path: &Path) -> bool {
    path.starts_with(storage::APPS)
        || path.starts_with(storage::USER)
        || path.starts_with("/tmp")
        || path.starts_with("/cache")
}

/// Check if path is a native app
pub fn is_native_app_path(path: &Path) -> bool {
    path.starts_with(storage::NATIVE_APPS)
}

/// Check if path is system-protected
pub fn is_system_path(path: &Path) -> bool {
    path.starts_with(storage::SYSTEM)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_app_paths() {
        let app_id = "my-app";
        assert_eq!(app::data_dir(app_id), PathBuf::from("/storage/apps/my-app/data"));
        assert_eq!(app::config_dir(app_id), PathBuf::from("/storage/apps/my-app/config"));
        assert_eq!(app::cache_dir(app_id), PathBuf::from("/cache/my-app"));
        assert_eq!(app::temp_dir(app_id), PathBuf::from("/tmp/my-app"));
    }

    #[test]
    fn test_userspace_detection() {
        assert!(is_userspace_path(Path::new("/storage/user/documents")));
        assert!(is_userspace_path(Path::new("/storage/apps/myapp")));
        assert!(is_userspace_path(Path::new("/tmp/test")));
        assert!(!is_userspace_path(Path::new("/storage/system/config")));
    }

    #[test]
    fn test_native_app_detection() {
        assert!(is_native_app_path(Path::new("/storage/native-apps/browser")));
        assert!(!is_native_app_path(Path::new("/storage/apps/myapp")));
    }
}

